### PR TITLE
AutoDiscover in Sharing Stage

### LINKED ISSUE #172: Implement AutoDiscovery for Sharing
Create a new script that is able to auto discover the server in a shared experience using the autodiscovery feature in the SharingClient.dll.

### PR DESCRIPTION
**Pull request by alexdrenea**
Added a flag in SharingStage to enable server auto discovery. 
Uses the first server that responds to connect to.
Fixes #172

### CODE REVIEW

File: Assets/HoloToolkit/Sharing/Scripts/SharingStage.cs [hunk 1]
```diff
@@ -24,45 +30,81 @@ public class SharingStage : MonoBehaviour
...
         private ConsoleLogWriter logWriter;
 
+        /// <summary>
+        /// Enables Server Discovery on the network
+        /// </summary>
+        private DiscoveryClient discoveryClient;
+        /// <summary>
+        /// Provides callbacks when server is discovered or lost.
+        /// </summary>
+        private DiscoveryClientAdapter listener;
```
Comment (NeerajW) [inline Assets/HoloToolkit/Sharing/Scripts/SharingStage.cs]: NIT: discoveryClientListener

File: Assets/HoloToolkit/Sharing/Scripts/SharingStage.cs [hunk 2]
```diff
@@ -24,45 +30,81 @@ public class SharingStage : MonoBehaviour
...
+        /// <summary>
+        /// Determines how often the discovery service should ping the network in search of a server.
+        /// </summary>
+        public float PingIntervalSec = 2;
```
Comment (NeerajW) [inline Assets/HoloToolkit/Sharing/Scripts/SharingStage.cs]: NIT: Convert comment to ToolTip please for better in editor usage.

File: Assets/HoloToolkit/Sharing/Scripts/SharingStage.cs [hunk 3]
```diff
@@ -24,45 +30,81 @@ public class SharingStage : MonoBehaviour
...
+        private float PingIntervalCurrent = 0;
```
Comment (NeerajW) [inline Assets/HoloToolkit/Sharing/Scripts/SharingStage.cs]: private variables should be: pingIntervalCurrent cased.

File: Assets/HoloToolkit/Sharing/Scripts/SharingStage.cs [hunk 4]
```diff
@@ -24,45 +30,81 @@ public class SharingStage : MonoBehaviour
...
+        private bool IsTryingToFindServer = false;
```
Comment (NeerajW) [inline Assets/HoloToolkit/Sharing/Scripts/SharingStage.cs]: isTryingToFindServer

File: Assets/HoloToolkit/Sharing/Scripts/SharingStage.cs [hunk 5]
```diff
@@ -80,6 +122,75 @@ private void OnDisable()
...
+            config.SetIsAudioEndpoint(this.IsAudioEndpoint);
+            config.SetLogWriter(this.logWriter);
+            config.SetServerAddress(this.ServerAddress);
+            config.SetServerPort(this.ServerPort);
+            config.SetProfilerEnabled(false);
+
+            this.sharingMgr = SharingManager.Create(config);
+
+            //delay sending notification so everything is initialized properly
+            Invoke("SendConnectedNotification", 1);
```
Comment (NeerajW) [inline Assets/HoloToolkit/Sharing/Scripts/SharingStage.cs]: Why do we need this? Is this an artifact of the new code?

**Comment by alexdrenea** [comment]
Fixed... Sorry for the delay.

**Comment by alexdrenea** [comment]
Any updates on this PR? I've been on an extended vacation (and will still be out in the coming week) but wanted to check in and see if there's anything missing.

**Comment by jwittner** [comment]
@alexdrenea I'll try to check this out tomorrow. Sorry for the long delay on this one.

**Comment by jwittner** [comment]
Worked great for me!

**Pull request merged**